Content.Server/GameObjects/EntitySystems/PowerSolarSystem.cs

```diff
@@ -11,55 +18,137 @@ namespace Content.Server.GameObjects.EntitySystems
-
-                // Total coverage calculated; apply it to the panel.
+                // There's supposed to be rotational logic here, but that implies putting it somewhere.
+                entity.Transform.WorldRotation = TargetPanelRotation;
+
                 var panel = entity.GetComponent<SolarPanelComponent>();
-                panel.Coverage = coverage;
+                if (panel.TimeOfNextCoverageUpdate < _gameTiming.CurTime)
+                {
+                    // Setup the next coverage check.
+                    TimeSpan future = SolarCoverageUpdateInterval + (SolarCoverageUpdateRandomInterval * _robustRandom.NextDouble());
+                    panel.TimeOfNextCoverageUpdate = _gameTiming.CurTime + future;
+                    UpdatePanelCoverage(panel);
+                }
+                TotalPanelPower += panel.Coverage * panel.MaxSupply;
             }
         }
+
+        private void UpdatePanelCoverage(SolarPanelComponent panel) {
+            IEntity entity = panel.Owner;
+
+            // So apparently, and yes, I *did* only find this out later,
+            // this is just a really fancy way of saying "Lambert's law of cosines".
+            // ...I still think this explaination makes more sense.
+
+            // In the 'sunRelative' coordinate system:
+            // the sun is considered to be an infinite distance directly up.
+            // this is the rotation of the panel relative to that.
+            // directly upwards (theta = 0) = coverage 1
+            // left/right 90 degrees (abs(theta) = (pi / 2)) = coverage 0
+            // directly downwards (abs(theta) = pi) = coverage -1
+            // as TowardsSun + = CCW,
+            // panelRelativeToSun should - = CW
+            var panelRelativeToSun = entity.Transform.WorldRotation - TowardsSun;
+            // essentially, given cos = X & sin = Y & Y is 'downwards',
+            // then for the first 90 degrees of rotation in either direction,
+            // this plots the lower-right quadrant of a circle.
+            // now basically assume a line going from the negated X/Y to there,
+            // and that's the hypothetical solar panel.
+            //
+            // since, again, the sun is considered to be an infinite distance upwards,
+            // this essentially means Cos(panelRelativeToSun) is half of the cross-section,
+            // and since the full cross-section has a max of 2, effectively-halving it is fine.
+            //
+            // as for when it goes negative, it only does that when (abs(theta) > pi)
+            // and that's expected behavior.
+            float coverage = (float)Math.Max(0, Math.Cos(panelRelativeToSun));
+
+            if (coverage > 0)
+            {
+                // Determine if the solar panel is occluded, and zero out coverage if so.
+                // FIXME: The "Opaque" collision group doesn't seem to work right now.
```

Comment:
What problems have you had with the `Opaque` flag?

Reply:
This is pretty much the source of why I thought it didn't do anything when I mentioned things in #1014 ; I didn't test aiming it at a window, so it just seemed like raycasts on Opaque didn't hit anything. Changing collision group was enough of a temporary fix so that testing other stuff would continue to work in the meantime.

Reply:
I don't understand - #1014 didn't actually have anything to do with Opaque not working as a flag. The issue was with the laser gun code, which was using the wrong kind of collision mask. Raycasts on Opaque definitely do collide with stuff.

Reply:
Now that you changed the flag back, you can remove this FIXME.